types/encoding.go

```diff
@@ -19,7 +19,17 @@ func (b Block) MarshalSia(w io.Writer) error {
 	w.Write(b.ParentID[:])
 	w.Write(b.Nonce[:])
 	encoding.WriteUint64(w, uint64(b.Timestamp))
-	return encoding.NewEncoder(w).EncodeAll(b.MinerPayouts, b.Transactions)
+	encoding.WriteInt(w, len(b.MinerPayouts))
+	for i := range b.MinerPayouts {
+		b.MinerPayouts[i].MarshalSia(w)
+	}
+	encoding.WriteInt(w, len(b.Transactions))
+	for i := range b.Transactions {
+		if err := b.Transactions[i].MarshalSia(w); err != nil {
+			return err
+		}
+	}
+	return nil
```

Comment:
Unfortunately we can't (easily) add sanity checks to `MarshalSia` methods. The obvious way to do it is to compare to `encoding.Marshal`, but `encoding.Marshal` will call the object's `MarshalSia` method if it's defined, so you get an infinite loop.

In addition, this method writes directly to `w`. In order to do a sanity check, we would need to write everything to a byte slice, compare it to the `encoding.Marshal` output, and then flush it to `w`. It's doable to but it negates some of the performance gain.